marshal_test.go

```diff
@@ -1504,6 +1504,87 @@ func BenchmarkSendOneRequest(b *testing.B) {
 	}
 }
 
+func TestUnconnectedSocket_fail(t *testing.T) {
+	withUnconnectedSocket(t, false)
+}
+
+func TestUnconnectedSocket_success(t *testing.T) {
+	withUnconnectedSocket(t, true)
+}
+
+func withUnconnectedSocket(t *testing.T, enable bool) {
+	srvr, err := net.ListenUDP("udp4", &net.UDPAddr{})
+	if err != nil {
+		t.Fatalf("udp4 error listening: %s", err)
+	}
+	defer srvr.Close()
+
+	x := &GoSNMP{
+		Version:                 Version2c,
+		Target:                  srvr.LocalAddr().(*net.UDPAddr).IP.String(),
+		Port:                    uint16(srvr.LocalAddr().(*net.UDPAddr).Port),
+		Timeout:                 time.Millisecond * 100,
+		Retries:                 2,
+		UseUnconnectedUDPSocket: enable,
+	}
+	if err := x.Connect(); err != nil {
+		t.Fatalf("error connecting: %s", err)
+	}
+
+	go func() {
+		buf := make([]byte, 256)
+		for {
+			n, addr, err := srvr.ReadFrom(buf)
+			if err != nil {
+				return
+			}
+			buf := buf[:n]
+
+			var reqPkt SnmpPacket
+			var cursor int
+			cursor, err = x.unmarshalHeader(buf, &reqPkt)
+			if err != nil {
+				t.Errorf("error: %s", err)
+			}
+			err = x.unmarshalPayload(buf, cursor, &reqPkt)
+			if err != nil {
+				t.Errorf("error: %s", err)
+			}
+
+			rspPkt := x.mkSnmpPacket(GetResponse, []SnmpPDU{
+				{
+					Name:  ".1.2",
+					Type:  Integer,
+					Value: 123,
+				},
+			}, 0, 0)
+			rspPkt.RequestID = reqPkt.RequestID
+			outBuf, err := rspPkt.marshalMsg()
+			if err != nil {
+				t.Errorf("ERR: %s", err)
+			}
+			// Temporary socket will use different source port, it's enough to break
+			// connected socket reply filters.
+			nsock, err := net.ListenUDP("udp4", nil)
+			if err != nil {
+				t.Errorf("can't create temporary reply socket: %v", err)
+			}
+			nsock.WriteTo(outBuf, addr)
+			nsock.Close()
+		}
+	}()
+
+	pdus := []SnmpPDU{{Name: ".1.2", Type: Null}}
+	reqPkt := x.mkSnmpPacket(GetResponse, pdus, 0, 0) //not actually a GetResponse, but we need something our test server can unmarshal
```

Comment:
Minor nit.

```suggestion
	// This is Not actually a GetResponse, but we need something our test server can unmarshall.
	reqPkt := x.mkSnmpPacket(GetResponse, pdus, 0, 0)
```

Reply:
I actually copied it from another test, together with comment. Shall we update the comment there as well?

Reply:
I don't typically ask to fix unrelated lines, but you're welcome to volunteer to do this cleanup for consistency if you agree with the change.